include/armadillo_bits/operator_times.hpp

```diff
@@ -538,29 +538,45 @@ operator*
   
   if( (pa.get_n_elem() > 0) && (pb.get_n_nonzero() > 0) )
     {
-    typename SpProxy<T2>::const_iterator_type y_it     = pb.begin();
-    typename SpProxy<T2>::const_iterator_type y_it_end = pb.end();
-    
-    const uword result_n_rows = result.n_rows;
-    
-    while(y_it != y_it_end)
+    #if defined(ARMA_USE_OPENMP) 
       {
-      const eT    y_it_val = (*y_it);
-      const uword y_it_col = y_it.col();
-      const uword y_it_row = y_it.row();
-      
-      for(uword row = 0; row < result_n_rows; ++row)
+      const uword Y_n_cols = pb.get_n_cols();
+      const uword n_threads = uword( mp_thread_limit::get() );
+
+      #pragma omp parallel for schedule(static) num_threads(n_threads)
+      for (uword i = 0; i < Y_n_cols; ++i)
         {
-        result.at(row, y_it_col) += pa.at(row, y_it_row) * y_it_val;
+        const uword col_ptr_1 = pb.get_col_ptrs()[i];
+        const uword col_ptr_2 = pb.get_col_ptrs()[i + 1];
+        const uvec idx = uvec(&pb.get_row_indices()[col_ptr_1], col_ptr_2 - col_ptr_1);
+        const Col<eT> y_csc_col = Col<eT>(&pb.get_values()[col_ptr_1], col_ptr_2 - col_ptr_1);
+        result.col(i) = x.cols(idx) * y_csc_col;
         }
+      }
+    #else
+      {
+      typename SpProxy<T2>::const_iterator_type y_it     = pb.begin();
+      typename SpProxy<T2>::const_iterator_type y_it_end = pb.end();
       
-      ++y_it;
+      const uword result_n_rows = result.n_rows;
+      
+      while(y_it != y_it_end)
+        {
+        const eT    y_it_val = (*y_it);
+        const uword y_it_col = y_it.col();
+        const uword y_it_row = y_it.row();
+        
+        for(uword row = 0; row < result_n_rows; ++row)
+          {
+          result.at(row, y_it_col) += pa.at(row, y_it_row) * y_it_val;
+          }
```

Comment:
I might suggest that it might be better to OpenMP-ize here instead.  I am a little concerned that the overhead of OpenMP will outweigh any speedup from the loop, but I am not sure.  I still think it is worth at least trying though.

Reply:
Thanks for suggestion Ryan. We had conversation via email with Conrad (accidentally I replied only to him). Conrad has done benchmarks and it seems there is significant overhead for certain cases where single-thread version is faster. (According to Conrad: preliminary tests it seems that the number of rows in the dense matrix must not be larger than about 1% of the number of columns). 
On my side I tested it with quite large matrices - dense had 8-256 rows and 1e4-3e5 columns.   Sparse had 1e4-3e5 rows and 1e5-5e6 columns with sparsity 0.999-0.9999. In my case speed-up was 3.5-3.8 on 4core/8thread machine.
I'm planning to set up more benchmarks in order to figure out some heuristic/tradeoff when to use openMP or not.

Reply:
I mean in my tests 99.9%-99.99% of elements were zeros and 0.01% -0.1% non zeros.

Reply:
Ah, got it, sorry for the misunderstanding.

Reply:
@rcurtin - I did a bunch of tests with parallellising the for loop as per your suggestion.  It always caused slowdowns :(  Column-wise addition is a relatively quick operation, so the openmp can't really do much here.